arch/lkl/kernel/signal.c

```diff
@@ -50,25 +61,100 @@ void lkl_process_trap(int signr, struct ucontext *uctx)
         if(signr == ksig.sig)
         {
+            LKL_TRACE("trap task %p %d\n", current, signr);
             handle_signal(&ksig, uctx);
             break;
         }
     }
 }
 
-void do_signal(struct pt_regs *regs)
+/*
+    Walk to the end of the list and make it point at the new node
+*/
+
+
+void append_ksignal_node(struct thread_info *task, struct ksignal_list_node* node)
 {
-    struct ksignal ksig;
-    struct ucontext uc;
+    volatile struct ksignal_list_node **node_ptr;
+    spin_lock(&task->signal_list_lock);
+    node_ptr = &task->signal_list;
 
-    LKL_TRACE("enter\n");
+    while (*node_ptr) {
+        volatile struct ksignal_list_node *next_node = *node_ptr;
+        node_ptr = &(next_node->next);
+    }
+        
+    *node_ptr = node;
+    spin_unlock(&task->signal_list_lock);
+}    
 
-    memset(&uc, 0, sizeof(uc));
-    initialize_uctx(&uc, regs);
+// probably needs to be called with the cpu lock
+void move_signals_to_task(void)
+{
+    struct ksignal ksig;    // place to hold retrieved signal
+    // get the lkl local version of the current task, so we can store the signals
+    // in a list hanging off it.
+    struct thread_info *current_task;
+    current_task = task_thread_info(current); 
     while (get_signal(&ksig)) {
-        LKL_TRACE("ksig.sig=%d", ksig.sig);
+<<<<<<< HEAD
+    	struct ksignal_list_node* node = kmalloc(sizeof(struct ksignal_list_node), GFP_KERNEL); // may sleep, is that ok?       
+        if (node == NULL) {
+            lkl_bug("kmalloc returned NULL");
+        }
+        LKL_TRACE("Appending task %p signal %d\n", current, ksig.sig);
+        memcpy(&node->sig, &ksig, sizeof(ksig));
+        node->next = NULL;
+        append_ksignal_node(current_task, node);
+    }
+}
+
+int get_next_ksignal(struct thread_info *task, struct ksignal* sig)
+{
+    struct ksignal_list_node *next;
+    struct ksignal_list_node *node;
+
+    spin_lock(&task->signal_list_lock);
+    node = task->signal_list;
+
+    if (!node) {
+        spin_unlock(&task->signal_list_lock);
+        return 0; // no pending signals
+    }
+
+    next = node->next;
+    task->signal_list = next; // drop the head
+    
+    spin_unlock(&task->signal_list_lock);
+    
+    memcpy(sig, &node->sig, sizeof(*sig)); // copy the signal back to the caller
+    LKL_TRACE("Fetching task %p signal %d\n", current, sig->sig);
+    kfree(node);
+    return 1;
+}    
+
+// Must be called without the cpu lock
+void send_current_signals(struct ucontext *uctx)
+{
+    struct thread_info *current_task = task_thread_info(current);
+    struct ksignal ksig;
+    
+    struct ucontext uc;
+    
+    if (uctx == NULL) {
+        uctx = &uc;
+        memset(uctx, 0, sizeof(uc));
+        initialize_uctx(uctx, NULL);
+    }
 
-        /* Whee!  Actually deliver the signal.  */
-        handle_signal(&ksig, &uc);
+    LKL_TRACE("enter\n");
+    
+    while (get_next_ksignal(current_task, &ksig)) {
+    	LKL_TRACE("ksig.sig=%d", ksig.sig);
```

Comment:
something seems off with the formatting here.

Reply:
Fixed.